crates/matrix-sdk/src/client/mod.rs

```diff
@@ -1538,6 +1815,37 @@ impl Client {
         request: Request,
         config: Option<RequestConfig>,
     ) -> HttpResult<Request::IncomingResponse>
+    where
+        Request: OutgoingRequest + Clone + Debug,
+        HttpError: From<FromHttpResponseError<Request::EndpointError>>,
+    {
+        let res = self.send_inner(request.clone(), config).await;
+
+        // If this is an `M_UNKNOWN_TOKEN` error and refresh token handling is active,
+        // try to refresh the token and retry the request.
+        if self.inner.handle_refresh_tokens {
+            if let Err(HttpError::Api(FromHttpResponseError::Server(ServerError::Known(
+                RumaApiError::ClientApi(error),
+            )))) = &res
+            {
+                if matches!(error.kind, ErrorKind::UnknownToken { .. }) {
+                    let refresh_res = self.refresh_access_token().await;
+
+                    if refresh_res.is_ok() {
```

Comment:
It's a good question. According to the spec there are only two errors possible when trying to refresh a token: the same `UnknownToken` error, or rate limiting.

So it's probably better to report the new error otherwise, if it's rate limiting, the user won't know that they can retry the request and they'll assume they have to login again.

Reply:
The issue is when there are several requests at the same time. For this to work properly we'd need `Error` to implement `Clone`, so we can pass the same error to the other requests, but it's not possible because errors outside of our crates don't implement it (eg `ReqwestError`).

We could treat the Matrix Client API errors differently because they implement `Clone` and it might be rate limiting, and the rest gets logged and we put a generic error.

Reply:
> We could treat the Matrix Client API errors differently because they implement Clone and it might be rate limiting, and the rest gets logged and we put a generic error.

sounds like a fair trade off.